fw_info structure change (#1378)
NordicPlayground/fw-nrfconnect-nrf: include/fw_info.rst

The purpose of the firmware information is to allow other images such as bootloaders, or infrastructure such as firmware servers, to gain information about the firmware image.

The firmware information structure has a 12-byte magic header and a verified binary layout to ensure that the format is portable and identifiable.

> It must be located at one of three offsets from the start of the image: 0x0, 0x200, 0x400, 0x800, 0x1000.

Reviewer: one of the following

Author: Thanks, fixed.

* The single, monotonically increasing version number of the image.
* The address through which to boot into the firmware (the vector table address).
  This address is not necessarily the start of the image.

> * A value that must be 0xFFFFFFFF.

Author: Can you elaborate? If this check is spoofed during validation, the signature check will still fail.

> * A value that must be 0xFFFFFFFF.

Reviewer 2: Since 0xFFFFFFFF is the  value of erased memory, my gut feeling is that it would be easier to spoof this value somehow, compared to a random value 0x2B68C12A

> * A value that must be 0xFFFFFFFF.

Author: Ok, I think I understand. For it to work in the OTP region, it would have to be something like 

    #define VALID_VAL 0x2B68FFFF
    #define INVALID_VAL 0x2B680000

> * A value that must be 0xFFFFFFFF.

Author: Implemented now. I also put the valid value in Kconfig.

> * A value that must be 0xFFFFFFFF.

Reviewer: Does this need a doc update?

> * A value that must be 0xFFFFFFFF.

Author: Yes, thanks. Updated docs now.

>   If this is set to 0 (e.g. by the bootloader), the image can quickly be established as invalid.

Author: In practice it needs only be different from `0xFFFFFFFF`.

>   If this is set to 0 (e.g. by the bootloader), the image can quickly be established as invalid.

Reviewer 2: If the code enforce that non `0xFFFFFFFF` implies invalid, then IMO we should state that in the docs, otherwise it could be considered a bug that `0xFFFFFFF0` is considered invalid.

>   If this is set to 0 (e.g. by the bootloader), the image can quickly be established as invalid.

Author: Updated docs

>   The bootloader will set this to 0 when the image fails validation in some way, so a costly validation doesn't need to be performed on every boot.

Reviewer: ```suggestion
  The bootloader will set this to 0 when the image fails validation in some way, so that there is no need to perform a costly validation on every boot.
```

>   The bootloader will set this to 0 when the image fails validation in some way, so that there is no need to perform a costly validation on every boot.

Reviewer 3: Maybe add a point on this value being guarded from denial-of-service invalidation by a third party

Author: Added a sentence

* The single, monotonically increasing version number of the image.
* The address through which to boot into the firmware (the vector table address).
  This address is not necessarily the start of the image.
* A value that can be modified in place to invalidate the firmware. See :option:`CONFIG_FW_INFO_VALID_VAL`.
  If this is set to any other value than :option:`CONFIG_FW_INFO_VALID_VAL` (e.g. by the bootloader), the image can quickly be established as invalid.
  The bootloader will set this to 0 when the image fails validation in some way, so that there is no need to perform a costly validation on every boot.

>   If the firmware is write protected before being booted by the bootloader, only the bootloader will be able to invalidate it.

Reviewer: ```suggestion
  If the firmware is write-protected before being booted by the bootloader, only the bootloader can invalidate it.
```

* EXT_API flags (32 individual bits for indicating the particulars of the EXT_API)
* EXT_API length (length of the following data)

An image's EXT_APIs are placed at the end of the information structure, immediately before its EXT_API requests.

An EXT_API request structure is a structure consiting of a description of the requested EXT_API in the form of an EXT_API header, plus a few more pieces of data.
The structure also contains the location of a RAM buffer into which a pointer to a matching EXT_API can be placed before booting.

> The RAM buffer must not be initialized, or otherwise touched during bootup of the image.

Reviewer 3: Maybe add a helpful hint, like __attribute__((noinit))

Author: Added a sentence.

This is to preserve the value written there before boot.
The EXT_API system allows a bootloader to parse the image's requests, the parse its (the bootloader's) own, or other images' EXT_APIs to find matches, then populate the RAM buffers with pointers to other images' EXT_APIs.

> The bootloader will in this way also be aware of whether all images have access to their dependencies.

Reviewer 2: Check this sentence.

To create an EXT_API, complete the following steps:

1. Create a unique ID for the EXT_API:

   .. code-block:: c

      #define MY_EXT_API_ID 0xBEEF

> #. Create a Kconfig entries using :file:`Kconfig.template.fw_info_ext_api`:

Reviewer: ```suggestion
#. Create Kconfig entries using :file:`Kconfig.template.fw_info_ext_api`:
```

Author: Done :)
